core/src/test/java/io/questdb/griffin/engine/table/AsyncFilteredRecordCursorFactoryTest.java

```diff
@@ -0,0 +1,113 @@
+/*******************************************************************************
+ *     ___                  _   ____  ____
+ *    / _ \ _   _  ___  ___| |_|  _ \| __ )
+ *   | | | | | | |/ _ \/ __| __| | | |  _ \
+ *   | |_| | |_| |  __/\__ \ |_| |_| | |_) |
+ *    \__\_\\__,_|\___||___/\__|____/|____/
+ *
+ *  Copyright (c) 2014-2019 Appsicle
+ *  Copyright (c) 2019-2022 QuestDB
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ ******************************************************************************/
+
+package io.questdb.griffin.engine.table;
+
+import io.questdb.Metrics;
+import io.questdb.cairo.O3Utils;
+import io.questdb.cairo.SqlJitMode;
+import io.questdb.cairo.sql.RecordCursorFactory;
+import io.questdb.cairo.sql.async.PageFrameReduceTask;
+import io.questdb.cairo.sql.async.PageFrameSequence;
+import io.questdb.griffin.AbstractGriffinTest;
+import io.questdb.mp.RingQueue;
+import io.questdb.mp.SCSequence;
+import io.questdb.mp.WorkerPool;
+import io.questdb.mp.WorkerPoolConfiguration;
+import io.questdb.std.Misc;
+import org.junit.Assert;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+public class AsyncFilteredRecordCursorFactoryTest extends AbstractGriffinTest {
```

Comment:
nit: would be nice to add some tests for query cancellation (early cursor close), navigation (`toTop()`) and timeout handling into this test (or `AsyncOffloadTest`). Otherwise, we test these scenarios as a part of the existing test suite only in single-threaded environment.

Reply:
test does exist: `IODispatcherTest#testJsonQueryWithInterruption`

Reply:
The query in that test may be changed in future, so that it no longer involves async offload and the test would still pass since it's not meant to be testing the code from this PR. Would be nice to at least add a comment there or introduce a separate test.

Reply:
I agree, it would be nice to add simpler test. This test was failing without circuit breaker in offload. Catching crash we had is a little harder. But i'll come back to it.